File: ts-plugins/src/hierarchy/hierarchy-generation-plugin.ts
import type { GenerationPlugin, PluginDescriptor } from '../api/types.js';
import type { HierarchicalFlowSpec } from './types.js';
import { HierarchyCodeGenerator } from './hierarchy-code-generator.js';

export class HierarchyGenerationPlugin implements GenerationPlugin<HierarchicalFlowSpec, Map<string, string>> {
  private readonly generator = new HierarchyCodeGenerator();

  descriptor(): PluginDescriptor {
    return {
      id: 'hierarchy',
      displayName: 'Hierarchy Generator',
      description: 'Compiles hierarchical authoring specs into flat TypeScript state config and builder skeleton.',
    };
  }
  kind() { return 'GENERATION' as const; }

  generate(input: HierarchicalFlowSpec): Map<string, string> {
    return new Map([
      [`${input.enumName}.ts`, this.generator.generateStateConfig(input)],
      [`${input.flowName}Generated.ts`, this.generator.generateBuilderSkeleton(input)],
    ]);
  }
}

File: java-plugins/src/main/java/org/unlaxer/tramli/plugins/hierarchy/HierarchyGenerationPlugin.java
package org.unlaxer.tramli.plugins.hierarchy;

import org.unlaxer.tramli.plugins.api.GenerationPlugin;
import org.unlaxer.tramli.plugins.api.PluginDescriptor;

import java.util.Map;

public final class HierarchyGenerationPlugin implements GenerationPlugin<HierarchicalFlowSpec, Map<String, String>> {
    private final String packageName;
    private final HierarchyCodeGenerator generator = new HierarchyCodeGenerator();

    public HierarchyGenerationPlugin(String packageName) {
        this.packageName = packageName;
    }

    @Override
    public PluginDescriptor descriptor() {
        return new PluginDescriptor("hierarchy", "Hierarchy Generator", "Compiles hierarchical authoring specs into flat Java enum and builder skeleton sources.");
    }

    @Override
    public Map<String, String> generate(HierarchicalFlowSpec input) {
        return Map.of(
                input.enumName() + ".java", generator.generateEnumSource(input, packageName),
                input.flowName() + "Generated.java", generator.generateBuilderSkeleton(input, packageName)
        );
    }
}
